vpr/src/place/net_cost_handler.cpp

```diff
@@ -0,0 +1,2277 @@
+ * @param pin_old_loc
+ * @param pin_new_loc
+ * @param src_pin
+ */
+static void update_bb(ClusterNetId net_id,
+                      t_bb& bb_edge_new,
+                      t_bb& bb_coord_new,
+                      vtr::NdMatrixProxy<int, 1> num_sink_pin_layer_new,
+                      t_physical_tile_loc pin_old_loc,
+                      t_physical_tile_loc pin_new_loc,
+                      bool src_pin);
+
+/**
+ * @brief Update the per-layer bounding box of "net_id" incrementally based on the old and new locations of the pin
+ * @param bb_edge_new
+ * @param bb_coord_new
+ * @param num_sink_pin_layer_new
+ * @param pin_old_loc
+ * @param pin_new_loc
+ * @param src_pin
+ */
+static void update_layer_bb(ClusterNetId net_id,
+                            std::vector<t_2D_bb>& bb_edge_new,
+                            std::vector<t_2D_bb>& bb_coord_new,
+                            vtr::NdMatrixProxy<int, 1> bb_pin_sink_count_new,
+                            t_physical_tile_loc pin_old_loc,
+                            t_physical_tile_loc pin_new_loc,
+                            bool is_output_pin);
+
+/**
+ * @brief This function is called in update_layer_bb to update the net's bounding box incrementally if
+ * the pin under consideration is not changing layer.
+ * @param net_id
+ * @param pin_old_loc
+ * @param pin_new_loc
+ * @param curr_bb_edge
+ * @param curr_bb_coord
+ * @param bb_pin_sink_count_new
+ * @param bb_edge_new
+ * @param bb_coord_new
+ */
+static inline void update_bb_same_layer(ClusterNetId net_id,
+                                        const t_physical_tile_loc& pin_old_loc,
+                                        const t_physical_tile_loc& pin_new_loc,
+                                        const std::vector<t_2D_bb>& curr_bb_edge,
+                                        const std::vector<t_2D_bb>& curr_bb_coord,
+                                        vtr::NdMatrixProxy<int, 1> bb_pin_sink_count_new,
+                                        std::vector<t_2D_bb>& bb_edge_new,
+                                        std::vector<t_2D_bb>& bb_coord_new);
+
+/**
+* @brief This function is called in update_layer_bb to update the net's bounding box incrementally if
+* the pin under consideration change layer.
+ * @param net_id
+ * @param pin_old_loc
+ * @param pin_new_loc
+ * @param curr_bb_edge
+ * @param curr_bb_coord
+ * @param bb_pin_sink_count_new
+ * @param bb_edge_new
+ * @param bb_coord_new
+ */
+static inline void update_bb_layer_changed(ClusterNetId net_id,
+                                           const t_physical_tile_loc& pin_old_loc,
+                                           const t_physical_tile_loc& pin_new_loc,
+                                           const std::vector<t_2D_bb>& curr_bb_edge,
+                                           const std::vector<t_2D_bb>& curr_bb_coord,
+                                           vtr::NdMatrixProxy<int, 1> bb_pin_sink_count_new,
```

Comment:
It will be updated, and as mentioned above, NdMatrixProxy is actually designed to be a lightweight class moved across functions and it only has pointers to the underlying data structure.

Reply:
OK.  Could try a ref, might be faster (although also might make pointer aliasing checks harder for the compiler, so good to check speed).